core/src/edm/DeviceMultiTrajectory.cpp

```diff
@@ -0,0 +1,165 @@
+    return static_cast<IndexType>(
+        m_tracks.measurements.at(m_tracks.states.measurement_index().at(istate))
+            .dimensions());
+}
+
+bool DeviceMultiTrajectory::has_impl(Acts::HashedString key, IndexType) const {
+
+    return hasColumn_impl(key);
+}
+
+bool DeviceMultiTrajectory::hasColumn_impl(Acts::HashedString key) const {
+
+    // Use the hashing literal from Acts.
+    using namespace Acts::HashedStringLiteral;
+
+    // Hard-code the available columns.
+    switch (key) {
+        case "chi2"_hash:
+        case "measdim"_hash:
+            return true;
+        default:
+            return false;
+    }
+}
+
+std::any DeviceMultiTrajectory::component_impl(Acts::HashedString key,
+                                               IndexType istate) const {
+
+    // Use the hashing literal from Acts.
+    using namespace Acts::HashedStringLiteral;
+
+    // Extract the requested component.
+    switch (key) {
+        case "chi2"_hash:
+            return &(m_tracks.states.smoothed_chi2().at(istate));
+        case "measdim"_hash:
+            return &(m_tracks.measurements
+                         .at(m_tracks.states.measurement_index().at(istate))
+                         .dimensions());
+        default:
+            throw std::runtime_error(
+                "traccc::edm::DeviceMultiTrajectory::component_impl no such "
+                "component: " +
+                std::to_string(key));
+    }
+}
+
+auto DeviceMultiTrajectory::parameters_impl(IndexType index) const
+    -> ConstTrackStateProxy::ConstParameters {
+
+    // Access the track state's parameters.
+    const bound_track_parameters<default_algebra>& deviceParams =
+        m_tracks.states.smoothed_params().at(index);
+
+    // Construct the parameters on the fly. Since Acts always expects the
+    // parameters to be FP64, and the device code may use FP32.
+    Acts::BoundVector actsParams;
+    actsParams << deviceParams.bound_local()[0], deviceParams.bound_local()[1],
+        deviceParams.phi(), deviceParams.theta(), deviceParams.qop(),
+        deviceParams.time();
+
+    // Construct the parameters expected by Acts.
+    return ConstTrackStateProxy::ConstParameters{actsParams.data()};
+}
+
+auto DeviceMultiTrajectory::covariance_impl(IndexType index) const
+    -> ConstTrackStateProxy::ConstCovariance {
+
+    // Access the track state's parameters.
+    const bound_track_parameters<default_algebra>& params =
+        m_tracks.states.smoothed_params().at(index);
+
+    // Construct the covariance on the fly.
+    Acts::BoundSquareMatrix covariance;
+    for (unsigned int i = 0; i < Acts::BoundSquareMatrix::RowsAtCompileTime;
+         ++i) {
+        for (unsigned int j = 0; j < Acts::BoundSquareMatrix::ColsAtCompileTime;
+             ++j) {
+            covariance(i, j) = params.covariance()[i][j];
+        }
+    }
+
+    // Construct the covariance expected by Acts.
+    return ConstTrackStateProxy::ConstCovariance{covariance.data()};
```

Comment:
So, this will be dangling immediately, right? This is a bit the weakness of the current design that I chose, where the interface expects to be able to hand out references that the caller can manipulate directly. This probably prevents doing on the fly conversion here.

Reply:
I didn't look at this in detail. Just assumed that the object probably makes a copy. 

It is a defendable design that it doesn't. But then indeed this code will need to be written differently. 🤔